File: server/routes/player.js
/**
 * Player Profile Routes
 * ======================
 * GET  /player/profile      - Get player profile with stats
 * PUT  /player/profile      - Update player profile
 * GET  /player/progress     - Get all game progress
 * GET  /player/achievements - Get achievements
 */

const express = require('express');
const db = require('../models/db');

const router = express.Router();

/**
 * Get full player profile with aggregated stats.
 */
router.get('/profile', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const playerResult = await db.query(
            'SELECT * FROM players WHERE id = $1 AND tenant_id = $2',
            [playerId, tenantId]
        );

        if (playerResult.rows.length === 0) {
            return res.status(404).json({ error: 'Player not found' });
        }

        const player = playerResult.rows[0];

        // Aggregated game stats
        const statsResult = await db.query(`
            SELECT
                COUNT(*) FILTER (WHERE play_count > 0) as games_started,
                COUNT(*) FILTER (WHERE stars >= 1) as games_completed,
                COUNT(*) FILTER (WHERE stars = 3) as games_mastered,
                COALESCE(SUM(stars), 0) as total_stars,
                COALESCE(SUM(high_score), 0) as combined_high_scores
            FROM game_progress
            WHERE player_id = $1 AND tenant_id = $2
        `, [playerId, tenantId]);

        const stats = statsResult.rows[0];

        res.json({
            playerId: player.id,
            displayName: player.display_name,
            avatarCharacter: player.avatar_character,
            isGuest: player.is_guest,
            totalScore: parseInt(player.total_score, 10),
            gamesPlayed: player.games_played,
            totalPlayTime: player.total_play_time,
            createdAt: player.created_at,
            lastLoginAt: player.last_login_at,
            stats: {
                gamesStarted: parseInt(stats.games_started, 10),
                gamesCompleted: parseInt(stats.games_completed, 10),
                gamesMastered: parseInt(stats.games_mastered, 10),
                totalStars: parseInt(stats.total_stars, 10),
                maxStars: 25 * 3, // 25 games × 3 stars
                combinedHighScores: parseInt(stats.combined_high_scores, 10)
            }
        });
    } catch (err) {
        next(err);
    }
});

/**
 * Update player profile (display name, avatar).
 */
router.put('/profile', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;
        const { displayName, avatarCharacter } = req.body;

        const updates = [];
        const values = [];
        let paramIndex = 3;

        if (displayName) {
            updates.push(`display_name = $${paramIndex++}`);
            values.push(displayName);
        }
        if (avatarCharacter) {
            updates.push(`avatar_character = $${paramIndex++}`);
            values.push(avatarCharacter);
        }

        if (updates.length === 0) {
            return res.status(400).json({ error: 'No updates provided' });
        }

        const result = await db.query(
            `UPDATE players SET ${updates.join(', ')} WHERE id = $1 AND tenant_id = $2 RETURNING *`,
            [playerId, tenantId, ...values]
        );

        const player = result.rows[0];
        res.json({
            playerId: player.id,
            displayName: player.display_name,
            avatarCharacter: player.avatar_character
        });
    } catch (err) {
        next(err);
    }
});

/**
 * Get all game progress for the player.
 */
router.get('/progress', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(
            'SELECT * FROM game_progress WHERE player_id = $1 AND tenant_id = $2',
            [playerId, tenantId]
        );

        const progress = {};
        for (const row of result.rows) {
            progress[row.game_id] = {
                highScore: parseInt(row.high_score, 10),
                bestTime: row.best_time ? parseInt(row.best_time, 10) : null,
                level: row.level,
                stars: row.stars,
                playCount: row.play_count,
                totalScore: parseInt(row.total_score, 10),
                lastPlayed: row.last_played_at
            };
        }

        res.json({ progress });
    } catch (err) {
        next(err);
    }
});

/**
 * Get player achievements.
 */
router.get('/achievements', async (req, res, next) => {
    try {
        const playerId = req.player.id;
        const tenantId = req.player.tenantId;

        const result = await db.query(
            'SELECT * FROM player_achievements WHERE player_id = $1 AND tenant_id = $2 ORDER BY earned_at DESC',
            [playerId, tenantId]
        );

        res.json({
            achievements: result.rows.map(row => ({
                id: row.achievement_id,
                earnedAt: row.earned_at,
                gameId: row.game_id
            }))
        });
    } catch (err) {
        next(err);
    }
});

module.exports = router;

File: server-rs/src/routes/player.rs
use axum::{extract::State, Json};
use serde_json::{json, Value};

use crate::error::AppResult;
use crate::middleware::auth::AuthPlayer;
use crate::middleware::tenant::TenantId;
use crate::models::player::*;
use crate::AppState;

pub async fn get_profile(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;

    let p: Player = sqlx::query_as("SELECT * FROM players WHERE id = $1 AND tenant_id = $2")
        .bind(player.id)
        .bind(tenant_id)
        .fetch_one(&state.db)
        .await?;

    // Aggregate stats
    let stats: (i64, i64, i64, Option<i64>) = sqlx::query_as(
        r#"SELECT
            COUNT(*)::bigint as games_started,
            COUNT(CASE WHEN play_count > 0 THEN 1 END)::bigint as games_completed,
            COUNT(CASE WHEN stars = 3 THEN 1 END)::bigint as games_mastered,
            SUM(stars)::bigint as total_stars
        FROM game_progress WHERE player_id = $1 AND tenant_id = $2"#,
    )
    .bind(player.id)
    .bind(tenant_id)
    .fetch_one(&state.db)
    .await?;

    Ok(Json(json!({
        "player": PlayerPublic::from(&p),
        "email": p.email,
        "isGuest": p.is_guest,
        "totalPlayTime": p.total_play_time,
        "lastLoginAt": p.last_login_at,
        "stats": {
            "gamesStarted": stats.0,
            "gamesCompleted": stats.1,
            "gamesMastered": stats.2,
            "totalStars": stats.3.unwrap_or(0),
        }
    })))
}

pub async fn update_profile(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
    Json(body): Json<ProfileUpdateRequest>,
) -> AppResult<Json<Value>> {
    let tenant_id = &tenant.0 .0;

    let mut updates = Vec::new();
    let mut params: Vec<String> = Vec::new();

    if let Some(ref name) = body.display_name {
        updates.push(format!("display_name = ${}", params.len() + 3));
        params.push(name.clone());
    }
    if let Some(ref avatar) = body.avatar_character {
        updates.push(format!("avatar_character = ${}", params.len() + 3));
        params.push(avatar.clone());
    }

    if updates.is_empty() {
        return Ok(Json(json!({"message": "No fields to update"})));
    }

    let sql = format!(
        "UPDATE players SET {} WHERE id = $1 AND tenant_id = $2 RETURNING id, display_name, avatar_character, total_score, games_played",
        updates.join(", ")
    );

    // Build dynamic query
    let mut query = sqlx::query_as::<_, (uuid::Uuid, String, String, i64, i32)>(&sql)
        .bind(player.id)
        .bind(tenant_id);

    for p in &params {
        query = query.bind(p);
    }

    let (id, name, avatar, score, played) = query.fetch_one(&state.db).await?;

    Ok(Json(json!({
        "player": {
            "playerId": id,
            "displayName": name,
            "avatarCharacter": avatar,
            "totalScore": score,
            "gamesPlayed": played,
        }
    })))
}

pub async fn get_all_progress(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(String, i64, Option<i32>, i32, i32, i32)> = sqlx::query_as(
        "SELECT game_id, high_score, best_time, stars, level, play_count FROM game_progress WHERE player_id = $1 AND tenant_id = $2",
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let mut progress = serde_json::Map::new();
    for (gid, hs, bt, stars, level, pc) in rows {
        progress.insert(
            gid,
            json!({
                "highScore": hs,
                "bestTime": bt,
                "stars": stars,
                "level": level,
                "playCount": pc,
            }),
        );
    }

    Ok(Json(json!({ "progress": progress })))
}

pub async fn get_achievements(
    State(state): State<AppState>,
    player: axum::Extension<AuthPlayer>,
    tenant: axum::Extension<TenantId>,
) -> AppResult<Json<Value>> {
    let rows: Vec<(String, Option<String>, chrono::DateTime<chrono::Utc>)> = sqlx::query_as(
        "SELECT pa.achievement_id, pa.game_id, pa.earned_at FROM player_achievements pa WHERE pa.player_id = $1 AND pa.tenant_id = $2",
    )
    .bind(player.id)
    .bind(&tenant.0 .0)
    .fetch_all(&state.db)
    .await?;

    let achievements: Vec<Value> = rows
        .iter()
        .map(|(aid, gid, earned)| {
            json!({"achievementId": aid, "gameId": gid, "earnedAt": earned})
        })
        .collect();

    Ok(Json(json!({ "achievements": achievements })))
}
